pkg/modifycontroller/modify_volume.go

```diff
@@ -53,59 +52,77 @@ func (ctrl *modifyController) modify(pvc *v1.PersistentVolumeClaim, pv *v1.Persi
 		return pvc, pv, delayModificationErr, false
 	}
 
-	if pvcSpecVacName != nil && curVacName == nil {
-		// First time adding VAC to a PVC
-		return ctrl.validateVACAndModifyVolumeWithTarget(pvc, pv)
-	} else if pvcSpecVacName != nil && curVacName != nil && *pvcSpecVacName != *curVacName {
-		// Check if PVC in uncertain state
-		_, inUncertainState := ctrl.uncertainPVCs[pvcKey]
-		if !inUncertainState {
-			klog.V(3).InfoS("previous operation on the PVC failed with a final error, retrying")
-			return ctrl.validateVACAndModifyVolumeWithTarget(pvc, pv)
-		} else {
-			vac, err := ctrl.vacLister.Get(*pvcSpecVacName)
-			if err != nil {
-				if apierrors.IsNotFound(err) {
-					ctrl.eventRecorder.Eventf(pvc, v1.EventTypeWarning, util.VolumeModifyFailed, "VAC "+*pvcSpecVacName+" does not exist.")
-				}
-				return pvc, pv, err, false
-			}
-			return ctrl.controllerModifyVolumeWithTarget(pvc, pv, vac, pvcSpecVacName)
+	pvcSpecVacName := ptr.Deref(pvc.Spec.VolumeAttributesClassName, "")
+	curVacName := ptr.Deref(pvc.Status.CurrentVolumeAttributesClassName, "")
+	status := pvc.Status.ModifyVolumeStatus
+
+	if status == nil && pvcSpecVacName == curVacName {
+		// No modification required, already reached target state
+		return pvc, pv, nil, false
+	}
+
+	if pvcSpecVacName == "" &&
+		(status == nil || status.Status != v1.PersistentVolumeClaimModifyVolumeInProgress) {
+		// User don't care the target state, and we've reached a relatively stable state. Just keep it here.
+		// Note: APIServer generally not allowing setting pvcSpecVacName to empty when curVacName is not empty.
+		klog.V(4).InfoS("stop reconcile for rolled back PVC", "PV", klog.KObj(pv))
+		// Don't try to revert anything here, because we only record the result of the last modification.
+		// We don't know what happened before. User can switch between InProgress/Infeasible/Pending status
+		// freely by modifying the spec.
+		return pvc, pv, nil, false
+	}
+
+	// Check if we should change our target
+	_, inUncertainState := ctrl.uncertainPVCs.Load(pvcKey)
+	if (status != nil && status.Status == v1.PersistentVolumeClaimModifyVolumeInProgress && inUncertainState) || pvcSpecVacName == "" {
```

Comment:
So I think there is a lot happening in this conditional, but some comments:

1. `status` has to be not nil for not `pvcSpec == ""` case and uncertain case. So may be, pull the `status != nil` check above.
2. Same as my comment in other conditional. I personally find these complex conditions hard to read in future unless I am really paying attention and having these complex conditions behind a function that returns boolean value makes this greatly more readable.
3.

Reply:
Refactored these branches. Hopes will make this more readable.

Reply:
I do not see that handled. My concern is - we are reading from `status.TargetVolumeAttributesClassName` and there is a very very small chance that `status` itself may be nil. We do not want external-resizer to crash if that happens.